Automatic installation of missing binaries

tmattio opened the pull request:
This PR adds a popup to install missing tools (`ocamllsp` and `ocamlformat` for now) when they are not found in the user's sandbox.

It installs the tools from source for now, which is time-consuming, so we'll want to use something like `opam-tools`, but I figured it was good enough for a first version, and we most likely want to use it as a fallback anyways.

Not ready for review yet, but I wanted to get some early feedback on the approach.

**TODO**

- [ ] ~~Cancellation~~
- [X] Display errors
- [ ] ~~What to do when the install command wants to upgrade/downgrade some dependencies? Should we use "-y"?~~
- [X] Do not try to run ocamllsp if missing (and avoid the won't restart error)
- [x] Use the toolchain corresponding to the sandboxe's ocaml compiler version
- [ ] What to do if the user does not have Opam? Can we install it for them (🙈)?
- [ ] What to do if the user does not have Opam but has Esy? Should we create a private sandbox with it?
- [x] Should we use the tools in the user's sandbox when they are available?

Fixes #294
Fixes #361
Fixes #353

ulugbekna commented on src/toolchain.ml:
```diff
@@ -416,23 +418,120 @@ let get_command toolchain bin args : Cmd.t =
+  let open Promise.Syntax in
+  let* missing_tools =
+    [ ("ocaml-lsp-server", get_lsp_command toolchain ~args:[ "--version" ])
+    ; ("ocamlformat", get_ocamlformat_command toolchain ~args:[ "--version" ])
```
Just a heads up regarding formatters and formatting:

You may have noticed that formatting is done by ocaml-lsp, not the extension. Moreover, `ocamllsp` invokes either `ocamlformat` or `refmt` based on the file that needs to be formatted. So it may make sense to handle logic for detecting whether a formatter is installed on the ocaml-lsp side. The lang server could send a notification if it can't find the required tool.

If you choose to do it on the extension side, keep in mind that a user may only need refmt installed or ocamlformat or both, so it's not straightforward.

ulugbekna replied:
Sorry my TODO comment was misleading

tmattio replied:
Makes sense - do you know if there's a specific notification for that purpose? I've tried to see if LSP had requests/notifications for this kind of things, but did not find anything.

tmattio commented:
So, there's a problem when trying to install dependencies in a switch that uses `ocaml-system`, as opposed to `ocaml-base-compiler`. That's because the commands won't inherit the global environment and thus `ocamlc` won't be in the path.

One solution is to install `ocaml-base-compiler` on behalf of the user when this happens, but then the tools installation takes **a lot** of time.

Another solution would be to use the global env when running opam commands, but I don't know if that's possible, and if it is, I'm not sure this is a good idea.

Any insights on this?

ulugbekna commented:
There isn't a specific notification for that purpose afaik, but you can
define custom notifications. Currently, custom notifications in ocaml-lsp
are called `Unknown_notification`s, iirc, and are `assert false`d instead
of being handled. I would recommend adding support for those, then define a
custom notif for missing tools.

The above taken into account, I would concentrate this PR on installing
missing ocaml-lsp and create another PR for other tools.
----

Sidenote: I haven't looked into the whole code and don't know what the
UX/UI look like, but another heads up: we need to handle installation in
both opam and esy.

It may make sense to start shaping up a common interface for `Opam` and
`Esy` modules with a fn `install`, for example.

On Wed, 2 Dec 2020 at 15:46, Thibaut Mattio <notifications@github.com>
wrote:

> *@tmattio* commented on this pull request.
> ------------------------------
>
> In src/toolchain.ml
> <https://github.com/ocamllabs/vscode-ocaml-platform/pull/463#discussion_r534070270>
> :
>
> >      Cmd.output command
>    in
>    match output with
> -  | Ok _ -> Ok ()
> -  | Error msg ->
> -    (* TODO: if ocamllsp not found, suggest to install it on press of a button;
> -       consider checking and suggesting installation for other tools: formatter, etc. *)
> -    Error (Printf.sprintf "Toolchain initialisation failed: %s" msg)
> +  | Ok _ -> None
> +  | Error _ -> Some binary
> +
> +let run_setup toolchain =
> +  let open Promise.Syntax in
> +  let* missing_tools =
> +    [ ("ocaml-lsp-server", get_lsp_command toolchain ~args:[ "--version" ])
> +    ; ("ocamlformat", get_ocamlformat_command toolchain ~args:[ "--version" ])
>
> Makes sense - do you know if there's a specific notification for that
> purpose? I've tried to see if LSP had requests/notifications for this kind
> of things, but did not find anything.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ocamllabs/vscode-ocaml-platform/pull/463#discussion_r534070270>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4YR6Y4MKI2XV4IQSDQ22LSSYLJLANCNFSM4UJQGA3A>
> .
>

tmattio commented:
> but you can define custom notifications

@rgrinberg I know you were wary of adding custom requests, how do you feel about this? 

> I would concentrate this PR on installing missing ocaml-lsp and create another PR for other tools

Good point. I'll try to handle it in this PR and if it ends up being cumbersome I'll split it into another one.

> It may make sense to start shaping up a common interface for `Opam` and `Esy` modules with a fn `install`, for example.

That's what this PR implements 🙂

tmattio commented:
@rgrinberg As discussed, I change the way we install the tools to install them in a private switch. It's working but needs a bit more work, in particular, to install in a private switch with the same OCaml version the users are using in their sandbox.

There are now too modules:

- `Sandbox`, for the users' project sandbox
- `Toolchain`, for the private switch containing OCaml Platform tools

The installation is **really** slow the first time, as expected, so we'll want to optimize that with pre-built binaries/`opam-tools` when possible.

rgrinberg commented:
> So, there's a problem when trying to install dependencies in a switch that uses ocaml-system, as opposed to ocaml-base-compiler. That's because the commands won't inherit the global environment and thus ocamlc won't be in the path.
Using the system compiler is out of question.

Using the system compiler is out of question anyway.

> I know you were wary of adding custom requests, how do you feel about this?

Very negatively. 

Would it be possible to move unrelated commits to separate PR's? It's quite hard to review the work otherwise.

> The installation is really slow the first time, as expected, so we'll want to optimize that with pre-built binaries/opam-tools when possible.

The other reason why I prefer to use opam-tools is that I'd like to add some second class support for non vscode users. It won't be done automatically, but at least they will need to run a binary that works around all the limitations.

mnxn commented:
> There are now two modules:
> 
>  - `Sandbox`, for the users' project sandbox
> 
> - `Toolchain`, for the private switch containing OCaml Platform tools

This terminology should be switched. "Sandbox" implies isolation of some kind, and `Global`/`Custom` are not isolated at all; they just contain a set of tools, a toolchain.

A "private switch containing OCaml Platform tools" fits the "Sandbox" label better since it's a toolchain isolated from the user's project.

> - What to do if the user does not have Opam? Can we install it for them (🙈)?
> - What to do if the user does not have Opam but has Esy? Should we create a private sandbox with it?

A user that has esy installed should not have to install opam to install the tools. This would be especially annoying on Windows, where the setup process for opam-repository-mingw is not very simple or fast.

I would expect that the extension would try to install ocamllsp in my current toolchain, using `opam install` or `esy add`.

rgrinberg commented on src/toolchain.mli:
```diff
@@ -1,85 +1,45 @@
-val equal : t -> t -> bool
+val setup_toolchain_sandbox :
+     unit
+  -> (unit, [> `Opam_not_available | `Switch_create_failed of string ]) result
```
Are we doing something specific for these errors? If not, then there's no need to introduce specific error types.

The old codebase suffered from this a problem: a massive amount of error types that all eventually ended up flushed down as strings. Let's not reintroduce it.

rgrinberg commented on .ocamlformat:
```diff
@@ -3,6 +3,8 @@ break-separators=before
 break-sequences=true
 doc-comments=before
+wrap-comments = true
+parse-docstrings = true
```
Should come in a separate PR.

rgrinberg replied:
Actually, we should just copy the config in the dune repo both here and in lsp. It will give us a consistent style.

rgrinberg commented on src/bindings/node/node.ml:
```diff
@@ -86,9 +86,15 @@ module ChildProcess = struct
+
+  let on_error t f = on t "error" f
+
+  val kill : t -> string -> bool [@@js.call]
```
Large binding changes should also be separate.

tmattio replied:
I reverted this, it's not needed anymore

rgrinberg commented:
> This terminology should be switched. "Sandbox" implies isolation of some kind, and Global/Custom are not isolated at all; they just contain a set of tools, a toolchain.

Global/Custom may indeed be isolated, but we just don't know their structure. For example they could come from a nix environment. I agree that this is confusing though.

> A user that has esy installed should not have to install opam to install the tools. This would be especially annoying on Windows, where the setup process for opam-repository-mingw is not very simple or fast.

Yes, that makes sense. Relying on opam is a probably a band aid until we get binary installs. However, I do not think that we should complicate the process by adding a separate pipeline for esy. The goal of this feature is to be transparent and reliable. We aren't going to get there if we need to deal with an additional package manager.

tmattio commented:
> I would expect that the extension would try to install ocamllsp in my current toolchain, using opam install or esy add.

That's how I implemented this first, but we reviewed the design with @rgrinberg and think it makes more sense to isolate the tools from the user's current sandbox.

One reason is to optimize the installation: if we install the tools in the sandbox every time, we waste the users' time for every new project they open

Another reason is that the tools have dependencies. These dependencies may conflict with the users' project, we shouldn't ask users to change their project dependencies because of the tooling.

Lastly, it'll be much simpler to manage and update the platform tools if we control the environment.

tmattio commented:
> This would be especially annoying on Windows, where the setup process for opam-repository-mingw is not very simple or fast.

I agree, it'd prefer not to have system dependencies as well and installing opam on Windows seems... We don't want to go there before 2.2...

At the same time I agree with @rgrinberg that this is a temporary solution until we get binaries, at which point it'll become a fallback solution. For Esy users who don't have opam, I think we can try to see if the tools are in the user sandbox if we cannot use the private toolchain (or the other way around).

tmattio commented:
> Would it be possible to move unrelated commits to separate PR's? It's quite hard to review the work otherwise.

I extracted the renaming of the `Toolchain` module in #466, but I rebased on top of it, since it depends on this change. Let's merge this PR first maybe to clean up the diff?

tmattio commented:
@rgrinberg @mnxn It works well for me, but as it is a critical part of the extension (if this does not work, the extension is unusable), would you mind testing on your side to make sure you're happy with everything?

mnxn commented:
Broken on windows:

![image](https://user-images.githubusercontent.com/25037249/101685142-3d7c8c80-3a1c-11eb-916c-e35f9a083c93.png)

<details>
<summary>Output channel log</summary>

```json
external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\opam.exe",
  "args": [
    "switch",
    "list",
    "-s"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "4.11.1+mingw64c\n",
    "stderr": ""
  }
}

1 switches
external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\opam.exe",
  "args": [
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "ocamlc",
    "--version"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "4.11.1\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\opam.exe",
  "args": [
    "exec",
    "--switch=vscode-ocaml-toolchain.4.11.1",
    "--",
    "true"
  ],
  "result": {
    "exitCode": 2,
    "stdout": "",
    "stderr": "[ERROR] The selected switch vscode-ocaml-toolchain.4.11.1 is not installed.\n"
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\opam.exe",
  "args": [
    "exec",
    "--switch=vscode-ocaml-toolchain.4.11.1",
    "--",
    "true"
  ],
  "result": {
    "exitCode": 2,
    "stdout": "",
    "stderr": "[ERROR] The selected switch vscode-ocaml-toolchain.4.11.1 is not installed.\n"
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\opam.exe",
  "args": [
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "which",
    "ocamlc"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "/home/Max/.opam/4.11.1+mingw64c/bin/ocamlc\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\opam.exe",
  "args": [
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "which",
    "ocamlformat"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "/home/Max/.opam/4.11.1+mingw64c/bin/ocamlformat\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\opam.exe",
  "args": [
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "which",
    "ocamllsp"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "/home/Max/.opam/4.11.1+mingw64c/bin/ocamllsp\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\opam.exe",
  "args": [
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "ocamlc",
    "--version"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "4.11.1\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\opam.exe",
  "args": [
    "switch",
    "create",
    "vscode-ocaml-toolchain.4.11.1",
    "ocaml-base-compiler.4.11.1"
  ],
  "result": {
    "exitCode": 5,
    "stdout": "",
    "stderr": "[ERROR] No compiler matching 'ocaml-base-compiler.4.11.1' found, use 'opam switch list-available' to see what is available, or use '--packages' to select packages explicitly.\n"
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\opam.exe",
  "args": [
    "install",
    "--switch=vscode-ocaml-toolchain.4.11.1",
    "-y",
    "ocaml-base-compiler",
    "ocaml-lsp-server",
    "ocamlformat"
  ],
  "result": {
    "exitCode": 2,
    "stdout": "",
    "stderr": "[ERROR] The selected switch vscode-ocaml-toolchain.4.11.1 is not installed.\n"
  }
}


```

</details>

The configured opam switch that I'm using does have all of the tools available. A few things I noticed:

- `opam exec --switch=vscode-ocaml-toolchain.4.11.1 -- true` is being run twice

- `which` is returning unix paths for Cygwin, which won't work because VSCode is running from normal Windows.

tmattio commented:
Thanks a lot for testing @mnxn! I'll fix these and will make sure it runs properly on Windows 🙂

rgrinberg commented:
Will try to get around to this this week. Might be delayed until next week though.

tmattio commented:
@mnxn 

> which is returning unix paths for Cygwin, which won't work because VSCode is running from normal Windows.

We're only considering the exit code of `which`, so this should still be ok, no?

tmattio commented on src/toolchain.ml:
```diff
@@ -0,0 +1,161 @@
+let compiler_of_version v =
+  match (Platform.t, Platform.arch) with
+  | Win32, X32 -> "ocaml-base-compiler." ^ v ^ "+mingw32c"
+  | Win32, _ -> "ocaml-base-compiler." ^ v ^ "+mingw64c"
```
@mnxn Do you know if there's going to be any issue using a 64 bits version of the compiler in a 32 bits Cygwin?

rgrinberg replied:
Before we worry about this, let's make sure there are even 32 bit users out there. I highly doubt there are any left.

mnxn replied:
Based on https://fdopen.github.io/opam-repository-mingw/installation/, it says "you can use [either archive] with either the 32-bit or 64-bit version of cygwin". The ocaml compiler isn't linked against cygwin1.dll either, so the architecture of cygwin shouldn't matter.

What does matter is the architecture of the system, which you handle how I expect here.

The only thing that might need changing is the prefix. On the same [installation page](https://fdopen.github.io/opam-repository-mingw/installation/), the documentation refers to `ocaml-variants` instead of `ocaml-base-compiler`.

mnxn commented:
> We're only considering the exit code of which, so this should still be ok, no?

Oh, if we're just using the exit codes that should be fine.

tmattio commented:
@mnxn This should work properly on Windows now, feel free to try again 🙂

mnxn commented:
Installation of the tools seems to work fine, but it installs them even if they're already available in my current switch.

<details>
<summary>Output channel log</summary>

```json
external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "switch",
    "list",
    "-s"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "4.11.1+mingw64c\n",
    "stderr": ""
  }
}

1 switches
external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "ocamlc",
    "--version"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "4.11.1\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "exec",
    "--switch=vscode-ocaml-toolchain.4.11.1",
    "--",
    "true"
  ],
  "result": {
    "exitCode": 2,
    "stdout": "",
    "stderr": "[ERROR] The selected switch vscode-ocaml-toolchain.4.11.1 is not installed.\n"
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "exec",
    "--switch=vscode-ocaml-toolchain.4.11.1",
    "--",
    "true"
  ],
  "result": {
    "exitCode": 2,
    "stdout": "",
    "stderr": "[ERROR] The selected switch vscode-ocaml-toolchain.4.11.1 is not installed.\n"
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "which",
    "ocamlformat"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "/home/Max/.opam/4.11.1+mingw64c/bin/ocamlformat\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "which",
    "ocamllsp"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "/home/Max/.opam/4.11.1+mingw64c/bin/ocamllsp\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "ocamlc",
    "--version"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "4.11.1\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "switch",
    "create",
    "vscode-ocaml-toolchain.4.11.1",
    "ocaml-variants.4.11.1+mingw64c"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "\n<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>\n[ocaml-variants.4.11.1+mingw64c] found in cache\n\n<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>\n-> installed base-bigarray.base\n-> installed base-threads.base\n-> installed base-unix.base\n-> installed ocaml-variants.4.11.1+mingw64c\n-> installed ocaml-config.1\n-> installed ocaml.4.11.1\nDone.\n# Run eval $(opam env) to update the current shell environment\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "install",
    "--switch=vscode-ocaml-toolchain.4.11.1",
    "-y",
    "ocaml-lsp-server",
    "ocamlformat"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "The following actions will be performed:\n  - install seq                     base     [required by tyxml]\n  - install conf-m4                 1        [required by ocamlfind]\n  - install ocamlbuild              0.14.0   [required by fpath, uuseg]\n  - install cmdliner                1.0.4    [required by ocamlformat]\n  - install dune                    2.7.1    [required by ocaml-lsp-server, ocamlformat]\n  - install ocamlfind               1.8.1    [required by ocaml-lsp-server]\n  - install uchar                   0.0.2    [required by uutf]\n  - install stdlib-shims            0.1.0    [required by ocaml-lsp-server]\n  - install sexplib0                v0.14.0  [required by base, ppxlib]\n  - install result                  1.5      [required by ocaml-lsp-server]\n  - install re                      1.9.0    [required by ocamlformat]\n  - install ppx_derivers            1.2.1    [required by ppxlib]\n  - install ocaml-version           3.0.0    [required by ocamlformat]\n  - install ocaml-migrate-parsetree 2.1.0    [required by ppxlib]\n  - install ocaml-compiler-libs     v0.12.3  [required by ppxlib]\n  - install menhirSdk               20201122 [required by menhir]\n  - install menhirLib               20201122 [required by menhir]\n  - install fix                     20201120 [required by ocamlformat]\n  - install easy-format             1.3.2    [required by yojson]\n  - install dune-build-info         2.7.1    [required by ocaml-lsp-server, ocamlformat]\n  - install cppo                    1.6.6    [required by odoc]\n  - install topkg                   1.0.3    [required by fpath, uuseg]\n  - install csexp                   1.3.2    [required by ocaml-lsp-server]\n  - install ppxlib                  0.20.0   [required by ocamlformat]\n  - install menhir                  20201122 [required by ocamlformat]\n  - install biniou                  1.2.1    [required by yojson]\n  - install uutf                    1.0.2    [required by ocamlformat]\n  - install astring                 0.8.5    [required by fpath, odoc]\n  - install dune-configurator       2.7.1    [required by base]\n  - install yojson                  1.7.0    [required by ocaml-lsp-server]\n  - install uucp                    13.0.0   [required by uuseg]\n  - install tyxml                   4.4.0    [required by odoc]\n  - install fpath                   0.7.3    [required by ocamlformat]\n  - install base                    v0.14.0  [required by ocamlformat]\n  - install ppx_yojson_conv_lib     v0.14.0  [required by ocaml-lsp-server]\n  - install dot-merlin-reader       3.4.2    [required by ocaml-lsp-server]\n  - install uuseg                   13.0.0   [required by ocamlformat]\n  - install odoc                    1.5.2    [required by ocamlformat]\n  - install stdio                   v0.14.0  [required by ocamlformat]\n  - install ocaml-lsp-server        1.3.0\n  - install ocamlformat             0.16.0\n===== 41 to install =====\n\n<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>\n[astring.0.8.5] found in cache\n[base.v0.14.0] found in cache\n[biniou.1.2.1] found in cache\n[cmdliner.1.0.4] found in cache\n[cppo.1.6.6] found in cache\n[csexp.1.3.2] found in cache\n[dot-merlin-reader.3.4.2] found in cache\n[dune.2.7.1] found in cache\n[dune-build-info.2.7.1] found in cache\n[dune-configurator.2.7.1] found in cache\n[easy-format.1.3.2] found in cache\n[fix.20201120] found in cache\n[fpath.0.7.3] found in cache\n[menhir.20201122] found in cache\n[menhirLib.20201122] found in cache\n[menhirSdk.20201122] found in cache\n[ocaml-compiler-libs.v0.12.3] found in cache\n[ocaml-lsp-server.1.3.0] found in cache\n[ocaml-migrate-parsetree.2.1.0] downloaded from https://github.com/ocaml-ppx/ocaml-migrate-parsetree/releases/download/v2.1.0/ocaml-migrate-parsetree-v2.1.0.tbz\n[ocaml-version.3.0.0] downloaded from https://github.com/ocurrent/ocaml-version/releases/download/v3.0.0/ocaml-version-v3.0.0.tbz\n[ocamlbuild.0.14.0] found in cache\n[ocamlfind.1.8.1] found in cache\n[odoc.1.5.2] found in cache\n[ocamlformat.0.16.0] downloaded from https://github.com/ocaml-ppx/ocamlformat/releases/download/0.16.0/ocamlformat-0.16.0.tbz\n[ppx_derivers.1.2.1] found in cache\n[ppx_yojson_conv_lib.v0.14.0] found in cache\n[re.1.9.0] found in cache\n[result.1.5] found in cache\n[sexplib0.v0.14.0] found in cache\n[stdio.v0.14.0] found in cache\n[stdlib-shims.0.1.0] found in cache\n[topkg.1.0.3] found in cache\n[tyxml.4.4.0] found in cache\n[ppxlib.0.20.0] downloaded from https://github.com/ocaml-ppx/ppxlib/releases/download/0.20.0/ppxlib-0.20.0.tbz\n[uchar.0.0.2] found in cache\n[uucp.13.0.0] found in cache\n[uuseg.13.0.0] found in cache\n[uutf.1.0.2] found in cache\n[yojson.1.7.0] found in cache\n\n<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>\n-> installed seq.base\n-> installed conf-m4.1\n-> installed cmdliner.1.0.4\n-> installed ocamlbuild.0.14.0\n-> installed ocamlfind.1.8.1\n-> installed uchar.0.0.2\n-> installed topkg.1.0.3\n-> installed uutf.1.0.2\n-> installed astring.0.8.5\n-> installed fpath.0.7.3\n-> installed dune.2.7.1\n-> installed cppo.1.6.6\n-> installed dune-build-info.2.7.1\n-> installed easy-format.1.3.2\n-> installed fix.20201120\n-> installed menhirLib.20201122\n-> installed menhirSdk.20201122\n-> installed ocaml-version.3.0.0\n-> installed ppx_derivers.1.2.1\n-> installed result.1.5\n-> installed sexplib0.v0.14.0\n-> installed stdlib-shims.0.1.0\n-> installed uucp.13.0.0\n-> installed re.1.9.0\n-> installed ocaml-migrate-parsetree.2.1.0\n-> installed ocaml-compiler-libs.v0.12.3\n-> installed csexp.1.3.2\n-> installed tyxml.4.4.0\n-> installed biniou.1.2.1\n-> installed uuseg.13.0.0\n-> installed dune-configurator.2.7.1\n-> installed yojson.1.7.0\n-> installed menhir.20201122\n-> installed odoc.1.5.2\n-> installed ppx_yojson_conv_lib.v0.14.0\n-> installed dot-merlin-reader.3.4.2\n-> installed ppxlib.0.20.0\n-> installed base.v0.14.0\n-> installed stdio.v0.14.0\n-> installed ocaml-lsp-server.1.3.0\n-> installed ocamlformat.0.16.0\nDone.\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "which",
    "ocamllsp"
  ],
  "result": {
    "exitCode": 0,
    "stdout": "/home/Max/.opam/4.11.1+mingw64c/bin/ocamllsp\n",
    "stderr": ""
  }
}

external command {
  "bin": "C:\\dev\\ocaml\\usr\\local\\bin\\ocaml-env.exe",
  "args": [
    "exec",
    "--",
    "opam",
    "exec",
    "--switch=4.11.1+mingw64c",
    "--",
    "ocamllsp"
  ]
}
````

</details>

In these logs, `which` is shown to return 0, but it still creates the new switch.

tmattio commented:
> Installation of the tools seems to work fine, but it installs them even if they're already available in my current switch.

That's the expected behavior, or at least the one I intended. The binaries used should still be the ones found in your switch, but the toolchain is set up regardless of the state of your sandbox. (since the toolchain is global to every project, and not specific to your current project).

I'm not sure if everyone agrees with the behavior though?

mnxn commented:
It's a bit weird user experience, since it takes 5 minutes to compile all the tools just to use what I already had available. 
I guess it's fine since it's a one-time setup. (We should make it possible to update the tools in this toolchain in a later PR)


A few more things I noticed:

- `vscode-ocaml-toolchain` doesn't show up in the toolchain selection:

| Before | After |
| - | - |
| ![before](https://user-images.githubusercontent.com/25037249/102533599-7dfe8a80-405a-11eb-8363-ceb2f528f4ab.png) | ![after](https://user-images.githubusercontent.com/25037249/102533654-8fe02d80-405a-11eb-8471-931c1051d052.png) |

Is this intentional?


- If I try to switch to a global sandbox (that doesn't have ocamlc avilable), the sandbox doesn't change and I get a `ERR spawn ocamlc ENOENT` in the developer console. I don't think the extension is checking whether `ocamlc` exists before doing `ocamlc --version`.

- `which` is not available on Windows unless the user has the cygwin executable on their PATH. We should use command prompt's `where` instead, which should work everywhere on windows (`cmd /c where`). This has the benefit of returning Windows paths in the output channel logs.

mnxn commented on src/sandbox.ml:
```diff
@@ -413,24 +414,29 @@ let get_command sandbox bin args : Cmd.t =
+  | Esy (esy, manifest) -> Some (Esy.exec esy ~manifest ~args:tools)
+  | _ -> None
+
+let ocaml_version sandbox =
+  let cmd = get_command sandbox "ocamlc" [ "--version" ] in
+  let open Promise.Result.Syntax in
+  let+ output = Cmd.output cmd in
+  String.strip output
```
Need a `Cmd.check` here.

mnxn commented on src/sandbox.ml:
```diff
@@ -413,24 +414,29 @@ let get_command sandbox bin args : Cmd.t =
-  get_command sandbox "ocamllsp" args
+let has_command sandbox bin =
+  let open Promise.Syntax in
+  let cmd = get_command sandbox "which" [ bin ] in
```
On Windows, this should be `get_command sandbox "cmd" [ "/c"; "where"; bin ]`.

tmattio commented:
> vscode-ocaml-toolchain doesn't show up in the toolchain selection: Is this intentional?

Yes, the toolchain switches should not be tampered with by the user, so not displaying it in the sandbox candidates mitigates the risk.

mnxn commented:
Looks good to me. Haven't encountered any other problems on Windows so far.

The README should probably be updated to describe the new workflow.

rgrinberg commented:
Hold off on merging for now. I'd like to test this out myself as well.

rgrinberg commented:
Any idea why CI failed on windows?

rgrinberg commented:
Just tested this out. Seems to be working smoothly. However, I have some concerns:

* These internal switches should not be visible to the user. Source based installation using opam is just an implementation detail.  Can we try using `OPAMROOT` to isolate our internal switches?

* We run the solver as part of the installation. That's quite fragile by itself and is incompatible with our plans for binary installation. We should have a fixed build plan for `os * ocaml` using lock files. And binary installation should use binaries that correspond to these build plans.

* We aren't handling upgrades. We should allow the user the option to upgrade ocamllsp when new versions are released.

tmattio commented:
> We should have a fixed build plan for os * ocaml using lock files.

I'm not sure how you would do that. AFAIK, opam can only install dependencies from lock files if they are in the filesystem, are you suggesting we store the lock file somewhere and execute the opam command where we've stored them?

rgrinberg commented:
> are you suggesting we store the lock file somewhere and execute the opam command where we've stored them?

Shipping such lock files with the plugin is certainly one option. It's not the only option, but it seems reasonable to me.

tmattio commented:
I have added support for upgrading the platform tools, but I now realize it is incompatible with providing a fixed build plan for the vscode switches (because upgrading will update to the latest dependencies, which we don't control).

A workaround would be to trigger the upgrade when the build plan for the switch has changed (this happens when we release a new version of the extension with an updated build plan). In which case we import the switch once again.

Alternatively, I'm in favor of not providing any build plan at all, even for the initial installation. IMO using Opam solver is solid enough considering that source installation will be the fallback and will be replaced with an installation of pre-built binaries as you mentioned.

rgrinberg commented:
I've thought over this and I'm still not sure what's the best course of action. On one hand, I hate having the solver decide the user's fat, on the other hand, I'm not 100% confident with making this work with lockfiles.

> A workaround would be to trigger the upgrade when the build plan for the switch has changed (this happens when we release a new version of the extension with an updated build plan). In which case we import the switch once again.

That sounds like the reasonable thing to do, and I don't see upgrades as the problem. The problem is that I imagine the whole process to generate lockfiles for all these combinations will be too unwieldy. The way I see it, we have 3 options:

1) Accept the PR as is.
2) Use lockfiles
3) Wait until binary installation.

@mnxn do you have a preference for any of these options?

tmattio commented:
I thought more about it too. There are two problems with the current approach:

- The platform tools may have dependency conflicts
- The switch creation takes way too long

I explored two alternatives:

- Reducing the switch creation time by caching the compiler (the same way Esy does). However, the first switch creation does not benefit from it, and that's the most important for us.
- Building the platform-tools with ocaml-ci, and downloading the binaries. However, ocaml-ci is not quite ready yet and we will have to provide a solution from source as a fallback anyway.

I discussed with @NathanReb and we thought one possible solution could be to make all of the platform-tools mono repositories (using `opam-monoreop`). In this scenario: 

- We wouldn't have to manage a separate opam switch, as there wouldn't be any dependency conflict.
- We could re-use the user's switch and install the tools in it if they are not already installed.
- We wouldn't need lock files, as the only dependency would be dune.

It seems to address all of the limitations of the other approaches while keeping the solution simple.

What do you think?

cc @emillon - Are we at a stage where we can release monorepos on opam?

rgrinberg commented:
I think using monorepo is fine, but we must be able to install it in the background ourselves (preferably without opam). If the user has to install this new tool manually, it defeats the purpose of automating the installation of lsp.

tmattio commented:
We discussed this PR with @avsm and think it can be merged as-is.

We agreed that there are some shortcomings in the OCaml Platform workflow that we need to address, but they will take time to put in place, and this PR still brings a lot of value for newcomers in the meantime.

@rgrinberg @mnxn I rebased and did some QA, but if you have some cycles to test this one last time, it would be great as it's a pretty invasive change. Also, let me know if you have any other concerns about the implementation.

mnxn commented on src/toolchain.ml:
```diff
@@ -0,0 +1,260 @@
+open Import
+
+let opamroot =
+  let home = Node.Os.homedir () in
```
I changed this to use Node's `os.homedir` because Windows doesn't use the `$HOME` environmental variables.

Before this change, the extension failed to activate on Windows.

tmattio replied:
I was sure Windows used `HOME`! Thanks @mnxn

mnxn commented on src/toolchain.ml:
```diff
@@ -0,0 +1,260 @@
+    in
+    Promise.return (Ok t)
+
+let to_pretty_string { version; _ } = version
```
I think it would be better if the pretty string had "OCaml" at the beginning so it's more apparent what extension the status bar item is coming from.
```suggestion
let to_pretty_string { version; _ } = "OCaml " ^ version
```

mnxn commented:
@tmattio:

I pushed a commit that uses the opam `--root` flag instead of the OPAMROOT environmental variable.

OPAMROOT didn't seem to work with Windows and `ocaml-env`, I always got errors like
```
[ERROR] Opam has not been initialised, please run `opam init'
```

I also changed the toolchain setup to wait for opam to finish installing the tools. Otherwise, the language server wouldn't start. Upgrading toolchain packages is still done asynchronously.

mnxn commented on src/toolchain.ml:
```diff
@@ -0,0 +1,240 @@
+  if exists then
+    Promise.Result.return ()
+  else
+    let+ result = Opam.init opam |> Cmd.output in
+    Result.map ~f:(fun _ -> ()) result
```
`opam init` takes a bit of time on the computers I tried it on. Maybe it should display progress too?

mnxn commented:
Just noticing this now, but I think the treeview buttons for removing an entire switch should require a confirmation of some kind. Same for removing a package.

Maybe in another PR?

tmattio commented:
Thanks @mnxn! I improved the workflow a bit to display progress for init and make sure that we start the server before we try to upgrade the toolchain (otherwise, opam exec seems to wait for the update to finish).

mnxn commented on src/toolchain.ml:
```diff
@@ -135,59 +127,71 @@ let get_install_command { opam = opam, switch; _ } tools =
+    Promise.Result.return ()
+  else
+    with_progress
+      ~msg:"Installing Platform Tools (this might take a few minutes)"
```
I think this should have a different message from `install_missing_tools` so the user has a little more information.

Maybe `Initializing Platform Toolchain (this might take a few minutes)`?

mnxn commented on vscode/interop/interop.ml:
```diff
@@ -117,6 +125,31 @@ module Js = struct
+    [@@js]
+
+    let t_of_js js_val =
+      match js_result_of_js js_val with
+      | { case = "ok"; ok; _ } -> Ok ([%js.to: Ok.t] ok)
+      | { case = "error"; error; _ } -> Error ([%js.to: Error.t] error)
+      | _ -> assert false
```
I changed this implementation because the previous version would reach the `assert false` when the `Ok` type is `unit`:
```
rejected promise not handled within 1 second: 0,248,Assert_failure,-10,0,vscode/interop/interop.ml,142,13
```

Because `unit` was turned into `undefined`, the function couldn't determine the difference between a unit value or the absence of one.

Checking a tag like I do here should avoid the problem.

mnxn reviewed (approved):
Works well for me. Good job!
